eval/topdown_test.go

```diff
@@ -0,0 +1,540 @@
+		{"ground: null", `p = true :- null = null`, "true"},
+		{"ground: array", `p = true :- [1,2,3] = [1,2,3]`, "true"},
+		{"ground: object", `p = true :- {"b": false, "a": [1,2,3]} = {"a": [1,2,3], "b": false}`, "true"},
+		{"ground: ref 1", `p = true :- a[2] = 3`, "true"},
+		{"ground: ref 2", `p = true :- b["v2"] = "goodbye"`, "true"},
+		{"ground: ref 3", `p = true :- d["e"] = ["bar", "baz"]`, "true"},
+		{"ground: ref 4", `p = true :- c[0].x[1] = c[0].z["q"]`, "true"},
+
+		// variables
+		{"var: a=b=c", "p[a] :- a = b, c = 42, b = c", "[42]"},
+		{"var: ref value", "p = true :- a[3] = x, x = 4", "true"},
+		{"var: ref values", "p = true :- a[i] = x, x = 2", "true"},
+		{"var: ref key", "p = true :- a[i] = 4, x = 3", "true"},
+		{"var: ref keys", "p = true :- a[i] = x, i = 2", "true"},
+		{"var: ref ground var", "p[x] :- i = 2, a[i] = x", "[3]"},
+		{"var: ref ref", "p[x] :- c[0].x[i] = c[0].z[j], x = [i, j]", `[[0, "p"], [1, "q"]]`},
+
+		// arrays and variables
+		{"pattern: array", "p[x] :- [1,x,3] = [1,2,3]", "[2]"},
+		{"pattern: array 2", "p[x] :- [[1,x],[3,4]] = [[1,2],[3,4]]", "[2]"},
+		{"pattern: array same var", "p[x] :- [2,x,3] = [x,2,3]", "[2]"},
+		{"pattern: array multiple vars", "p[z] :- [1,x,y] = [1,2,3], z = [x, y]", "[[2, 3]]"},
+		{"pattern: array multiple vars 2", "p[z] :- [1,x,3] = [y,2,3], z = [x, y]", "[[2, 1]]"},
+		{"pattern: array ref", "p[x] :- [1,2,3,x] = [a[0], a[1], a[2], a[3]]", "[4]"},
+		{"pattern: array = ref", "p[x] :- [true, false, x] = c[i][j]", `["foo"]`},
+		{"pattern: array = ref (reversed)", "p[x] :-  c[i][j] = [true, false, x]", `["foo"]`},
+		{"pattern: array = var", "p[y] :- [1,2,x] = y, x = 3", "[[1,2,3]]"},
+
+		// objects and variables
+		{"pattern: object val", `p[y] :- {"x": y} = {"x": "y"}`, `["y"]`},
+		{"pattern: var key error 1", `p[x] :- {x: "y"} = {"x": "y"}`, fmt.Errorf("cannot unify object with variable key: x")},
+		{"pattern: var key error 2", `p[x] :- {"x": "y"} = {x: "y"}`, fmt.Errorf("cannot unify object with variable key: x")},
+		{"pattern: var key error 3", `p = true :- {"x": [{y: "z"}]} = {"x": [{"y": "z"}]}`, fmt.Errorf("cannot unify object with variable key: y")},
+		{"pattern: object same var", `p[x] :- {"x": x, "y": x} = {"x": 1, "y": 1}`, "[1]"},
+		{"pattern: object multiple vars", `p[z] :- {"x": x, "y": y} = {"x": 1, "y": 2}, z = [x, y]`, "[[1, 2]]"},
+		{"pattern: object multiple vars 2", `p[z] :- {"x": x, "y": 2} = {"x": 1, "y": y}, z = [x, y]`, "[[1, 2]]"},
+		{"pattern: object ref", `p[x] :- {"p": c[0].x[0], "q": x} = c[i][j]`, `[false]`},
+		{"pattern: object = ref", `t[x] :- {"p": p, "q": q} = c[i][j], x = [i, j, p, q]`, `[[0, "z", true, false]]`},
+		{"pattern: object = ref (reversed)", `t[x] :- c[i][j] = {"p": p, "q": q}, x = [i, j, p, q]`, `[[0, "z", true, false]]`},
+		{"pattern: object = var", `p[x] :- {"a": 1, "b": b} = x, b = 2`, `[{"a": 1, "b": 2}]`},
+		{"pattern: object/array nested", `p[ys] :- f[i] = {"xs": [2.0], "ys": ys}`, `[[3.0]]`},
```

Comment:
Can always use more nested tests that combine different types, e.g.
p[x] = y :- f[i] = {"xs": [x], "ys": [y]}